### PR TITLE
KAFKA-15803: Update leader epoch in commitAsync and committed

### PR DESCRIPTION
**Pull request by lucasbru**
To align the new consumer completely with the legacy consumer,
we need to update the latest seen leader epoch in the metadata
both when we commit offsets with a leader epoch using
`commitAsync` and when we fetch the latest committed offsets
using `committed`.

We add unit tests to `AsyncKafkaConsumer` to test that
metadata is correctly updated. We also add a corresponding test
for `commitSync`, which already had the leader epoch logic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by AndrewJSchofield — APPROVED** [review]
This PR looks good to me.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java [hunk 1]
```diff
@@ -204,6 +206,24 @@ public void testCommitted() {
...
+        HashMap<TopicPartition, OffsetAndMetadata> topicPartitionOffsets = new HashMap<>();
+        topicPartitionOffsets.put(t0, new OffsetAndMetadata(10L, Optional.of(2), ""));
+        topicPartitionOffsets.put(t1, null);
+
+        CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> committedFuture = new CompletableFuture<>();
+        committedFuture.complete(topicPartitionOffsets);
+
+        try (MockedConstruction<OffsetFetchApplicationEvent> ignored = offsetFetchEventMocker(committedFuture)) {
+            assertDoesNotThrow(() -> consumer.committed(topicPartitionOffsets.keySet(), Duration.ofMillis(1000)));
+            verify(testBuilder.metadata).updateLastSeenEpochIfNewer(t0, 2);
```
Comment (cadonna) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: Could you please test multiple  calls to `updateLastSeenEpochIfNewer()`?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java [hunk 2]
```diff
@@ -204,6 +206,24 @@ public void testCommitted() {
...
+            verify(applicationEventHandler).add(ArgumentMatchers.isA(OffsetFetchApplicationEvent.class));
```
Comment (cadonna) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: Is there a reason you put these two verifications inside the try-resource-block but not in the other tests?

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @lucasbru !

I have a couple of minor comments

**Review by lucasbru — COMMENTED** [review]
Thanks, @cadonna and @AndrewJSchofield. I updated the PR according to the comments.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java (hunk 2 above)
Comment (lucasbru) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: Done

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java (hunk 1 above)
Comment (lucasbru) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: Done

**Review by cadonna — APPROVED** [review]
Thanks for the updates!

LGTM!

**Comment by lucasbru** [comment]
Test failures unrelated

**Pull request merged**